### PR TITLE
DO NOT MERGE: reference implementation configurable TLS postgres

### PR DESCRIPTION
**Pull request by doodlesbykumbi**
Resolves https://github.com/cyberark/secretless-broker/issues/463

the idea is
+ parse backend connection options
+ dial backend to get net.Conn
+ check if the connections options for the backend service specify tls and how
+ if tls; check if the server supports ssl, fail if not
+ if tls; upgrade net.Conn to tls.Conn with relevant tls config, given backend connection options
+ business as usual

default tls behavior is `required`

References:

https://github.com/lib/pq/blob/master/ssl.go
https://github.com/CrunchyData/crunchy-proxy/tree/master/connect

for mysql https://github.com/go-sql-driver/mysql/blob/master/packets.go#L322-L335

### CODE REVIEW

**Comment by izgeri** [comment]
@doodlesbykumbi which issue is this meant to help resolve? Is it #460?

File: test/pg_handler/etc/cert.sh [hunk 1]
```diff
@@ -0,0 +1,62 @@
+#!/bin/bash
+
+BASEDIR=$(dirname "$0")
+LOGGING_PREFIX="gen_cert.sh >> "
+
```
Comment (jonahx) [inline test/pg_handler/etc/cert.sh]: I wonder if we should be re-using something standard here.  eg, there are:

- https://github.com/conjurdemos/conjur-intro/tree/master/tools/simple-certificates
- https://github.com/conjurinc/appliance/tree/master/features/support/ca
- https://github.com/cyberark/conjur/blob/master/cucumber/api/features/support/ca_helpers.rb

we seem to be duplicating a lot of work around certs as a team...

File: test/pg_handler/etc/cert.sh [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+echo "${LOGGING_PREFIX} Creating server certificate"
+openssl req -new -key ${BASEDIR}/server.key -out ${BASEDIR}/server.csr -subj "/C=UK/ST=/L=/O=IBM/OU=AIOS/CN=pg" -reqexts SAN -config <(cat /etc/ssl/openssl.cnf <(printf "[SAN]\nsubjectAltName=DNS:pg,DNS:localhost"))
+echo "${LOGGING_PREFIX} Server certificate signing request (${BASEDIR}/server.csr) is:"
+openssl req -verify -in ${BASEDIR}/server.csr -text -noout
+echo
+
+# use our CA certificate and key to create a signed version of the server certificate
+echo
+echo "${LOGGING_PREFIX} Signing server certificate using our root CA certificate and key"
+openssl x509 -req -sha256 -days 365 -in ${BASEDIR}/server.csr -CA ${BASEDIR}/rootCA.crt -CAkey ${BASEDIR}/rootCA.key -CAcreateserial -out ${BASEDIR}/server.crt -extensions SAN -extfile <(cat /etc/ssl/openssl.cnf <(printf "[SAN]\nsubjectAltName=DNS:pg,DNS:localhost"))
```
Comment (jonahx) [inline test/pg_handler/etc/cert.sh]: the `<()` around `printf` shouldn't be needed.  can just do...

```
<(cat /etc/ssl/openssl.cnf; printf "[SAN]\nsubjectAltName=DNS:pg,DNS:localhost")
```

**Comment by doodlesbykumbi** [comment]
@izgeri this PR came out of #460 but isn't the agreed upon deliverable (a plan). This was meant to be a proof of concept more than anything but it's ended up morphing into an actual implementation.

File: test/pg_handler/etc/cert.sh (hunk 1 above)
Comment (doodlesbykumbi) [reply test/pg_handler/etc/cert.sh]: agreed! in fact this whole script is not necessary. we only need the certs. those could be generated once and checked into source control.

**Comment by izgeri** [comment]
Closing this PR, as content is being ported into PRs for #533 and #535

**Comment by izgeri** [comment]
leaving the branch open for now in case content still needs to be grabbed from here. @doodlesbykumbi @jonahx please delete the branch once you're done with it

**Pull request closed without merging**